### ISSUE TITLE
Is the logical pixel size 150 ppi instead of 96 ppi?

### ISSUE THREAD

**Issue by josvos** [flutter/flutter#22963]
On https://docs.flutter.io/flutter/dart-ui/Window/devicePixelRatio.html it is mentioned that there are typically 96 logical pixels per inch.  But in practice I see about 150 logical pixels per inch.  For example, a Galaxy Tab A `MediaQuery` returns 1280x800 (physically it is 1920x1200, the `devicePixelRatio` is reported correctly at 1.5).  Its longest side is about 8.5" with 1280 logical pixels.  Also, drawing fixed size boxes show the same pixel ratios.  The Nexus 10 tablet has similar logical figures (`devicePixelRatio` is there 2.0, which is correct).

Is the above referred documentation wrong?

**Comment by josvos** [comment]
After reading some literature, I found that the Android base pixel density seems to be 160 dpi (also called *MDPI* or the *baseline density*).  It looks like the Flutter values (e.g. via `MediaQuery`) are the same, although most devices tend to have approx. 150 (logical) ppi instead of 160.

**Comment by cubuspl42** [comment]
A very good question, my measurements are similar.

**Comment by esDotDev** [comment]
Would it be possible to get some clarification on why the docs list 96px, but measurements point to 160px or so?

**Comment by esDotDev** [comment]
It's been a year and a half now since this was originally logged, and still no basic explanation on why the docs are stating 96, when it's actually ~150 on mobile.

To complicate things, on the desktop and web, it does appear to be properly using 96. That means you need to check which platform you are on to get the correct size, but you can't include the Plaforms.is checks, because they are part of the .io package.

So basically there is now, no robust way, to figure out how many inches your layout will occupy.

**Comment by AissamOuajib** [comment]
Hey, I'm developing a Ring Sizer app and I need the measures to exact across all devices, but I encountered this problem, some devices work with 150 and other with 160 and on one device I tested I found  that 167 is what works. so is there any solution to that? 
Is there a way to measure the exact ppi of the device so I can use it to get the exact ring size in mm?

**Comment by bugeats** [comment]
I'm hitting this snag too. What I really need is what @AissamOuajib is asking for: a cross-platform abstraction to set dimensions in millimeters. Something like so:

```dart
SizedBox(
  width: mm(5),
  height: mm(9.5),
)
```

**Comment by esDotDev** [comment]
This should be what logical pixels already is. If it was reliably 100px / Inch, that would be 4px/mm. 

It's just that the DPI used is not consistent,

**Comment by AissamOuajib** [comment]
Hey, @bugeats I solved the issue by using this package: device_ext_info to get the right ppi of the device just like this:

Future<double> _getPPI(double screenDiagonal) async {
    final screenSizeInches =  await DeviceExtInfo.screenSizeInches;
    print(screenSizeInches);
    return screenDiagonal/screenSizeInches;
 }

and got the screen diagonal like this: 

final width = MediaQuery.of(context).size.width;
final height = MediaQuery.of(context).size.height;
final diagonal = sqrt(width*width+height*height);

and then all I do to transform the logical flutter pixel to mm is multiply by the ppi/25.4(ppi is what I get from the function above), note that I needed to use a FutureBuilder cause the function is async.

Now it works on every device, at least on all the ones I tested and I tested a bunch of them with deferent sizes and for both Android and iOS!

**Comment by esDotDev** [comment]
Did you test desktop and web as well?

**Comment by AissamOuajib** [comment]
Hey @esDotDev, No I didn't, I only tested on Android and iOS real devices.
The Ring Sizer app I'm working on is aimed only for mobile, you can test it yourself and let me know if it works on Desktop and Web as well!

**Comment by amir1061** [comment]
> Hey, I'm developing a Ring Sizer app and I need the measures to exact across all devices, but I encountered this problem, some devices work with 150 and other with 160 and on one device I tested I found that 167 is what works. so is there any solution to that? Is there a way to measure the exact ppi of the device so I can use it to get the exact ring size in mm?



> Hey, @bugeats I solved the issue by using this package: device_ext_info to get the right ppi of the device just like this:
> 
> ```
> Future<double> _getPPI(double screenDiagonal) async {
>     final screenSizeInches =  await DeviceExtInfo.screenSizeInches;
>     print(screenSizeInches);
>     return screenDiagonal/screenSizeInches;
>  }
> ```
> 
> and got the screen diagonal like this:
> 
> ```
> final width = MediaQuery.of(context).size.width;
> final height = MediaQuery.of(context).size.height;
> final diagonal = sqrt(width*width+height*height);
> ```
> 
> and then all I do to transform the logical flutter pixel to mm is multiply by the `ppi/25.4`(ppi is what I get from the function above), note that I needed to use a FutureBuilder cause the function is async.
> 
> Now it works on every device, at least on all the ones I tested and I tested a bunch of them with deferent sizes and for both Android and iOS!

Hi @aissamouajib ....I'm developing the same app can you please help me with the px to mm logic and how the pakage you mentioned works? Your help would be really helpful and appreciatable. thanks

**Comment by PuzzleTakX** [comment]
Use this

https://pub.dev/packages/ring_size_view

**Comment by nukeolay** [comment]
You can try this plugin [https://pub.dev/packages/display_metrics]. It retrieves the physical screen size in inches, the PPI, and can convert millimeters and inches to Flutter logical pixels and vice versa. It started with Android and iOS support, and now also works on Windows, macOS, and Web.


If you need accurate screen dimensions, PPI, or unit conversions across platforms, this might be helpful.
